### ISSUE TITLE
sidecar on aws requires spring.main.allow-bean-definition-override

### ISSUE THREAD

**Issue by gadamsciv** [spring-cloud/spring-cloud-netflix#3545]
I'm running spring-cloud-netflix sidecar on aws, and have recently upgraded from spring-boot 2.0.6 to 2.1.5 and spring-cloud Finchley.SR2 to Greenwich.SR1. 

Per spring-cloud documentation https://cloud.spring.io/spring-cloud-static/Greenwich.SR1/single/spring-cloud.html#_using_eureka_on_aws when running a eureka client on aws, one should customize the EurekaInstanceConfigBean to populate dataCenterInfo using AmazonInfo. This works fine for standalone eureka clients, but SidecarConfiguration.EurekaInstanceConfigBeanConfiguration is protected and creates a EurekaInstanceConfigBean with @Bean, and lacks a way to inject customization that would set dataCenterInfo. 

The only way I've been able to get this working is to set spring.main.allow-bean-definition-overriding to true in bootstrap.yml and continue creating my own EurekaInstanceConfigBean, overriding the bean created in SidecarConfiguration. This feels like a bad approach, since it seems like bean overriding is viewed as a bad thing. 

Could some way to customize the EurekaInstanceConfigBean without using bean overriding be added to sidecar?

**Comment by spencergibb** [comment]
Looks like the sidecar eureka instance config bean could use a conditional on missing bean annotation. PRs welcome.

**Comment by gadams00** [comment]
I've attempted adding @ConditionalOnMissingBean to the SidecarConfiguration.EurekaInstanceConfigBeanConfiguration.eurekaInstanceConfigBean method, but I'm still getting the BeanDefinitionOverrideException. If I reverse things and add @ConditionalOnMissingBean to my user-supplied EurekaInstanceConfigBean, that bean is omitted and the exception never happens. I must have some kind of ordering issue.

**Comment by spencergibb** [comment]
The other thing that likely has to happen is similar to what happened with `@EnableEurekaServer`. Rather than importing an `@Configuration` directly with all the beans, it imports one that creates a marker bean, then the original `@Configuration` is moved to auto-configuration, then you can apply ordering.

**Comment by gadams00** [comment]
I've got a fork over here: https://github.com/gadams00/spring-cloud-netflix

I've committed changes that I think you're talking about, but SidecarApplicationTests are failing. This is a little outside my comfort zone, but I think it's because SidecarAutoConfiguration.EurekaInstanceConfigBeanConfiguration.eurekaInstanceConfigBean is not ConditionalOnMissingBean, and EurekaClientAutoConfiguration is creating one by default. Before making the sidecar EurekaInstanceConfigBean conditional, it was always overriding the default created by eureka client auto config.

I guess what I need is some way to conditionally override eureka client's instance config bean (to inject standard SidecarProperties info), but allow applications to create their own instance config bean, and if that's present use that instead. I'm not sure how best to proceed with that.
